### PR TITLE
Cleanup macro-expanded code in `rustc_type_ir`

### PR DESCRIPTION
**Pull request by WaffleLapkin**
We could of course just leave this as-is, but every time I go-to-def to this file it's painful to see all this `(&A(ref __self_1_0),)` stuff.

### CODE REVIEW

File: compiler/rustc_type_ir/src/sty.rs [hunk 1]
```diff
@@ -301,61 +301,37 @@ impl<I: Interner> Clone for TyKind<I> {
...
+                (GeneratorWitness(a_g), GeneratorWitness(b_g)) => a_g == b_g,
+                (Tuple(a_t), Tuple(b_t)) => a_t == b_t,
+                (Projection(a_p), Projection(b_p)) => a_p == b_p,
+                (Opaque(a_d, a_s), Opaque(b_d, b_s)) => a_d == b_d && a_s == b_s,
+                (Param(a_p), Param(b_p)) => a_p == b_p,
+                (Bound(a_d, a_b), Bound(b_d, b_b)) => a_d == b_d && a_b == b_b,
+                (Placeholder(a_p), Placeholder(b_p)) => a_p == b_p,
+                (Infer(a_t), Infer(b_t)) => a_t == b_t,
+                (Error(a_e), Error(b_e)) => a_e == b_e,
+                _ => true, // unreachable
```
Comment (oli-obk) [inline compiler/rustc_type_ir/src/sty.rs:333]: Make it a panic in debug mode?

File: compiler/rustc_type_ir/src/sty.rs [hunk 2]
```diff
@@ -366,259 +342,145 @@ impl<I: Interner> Eq for TyKind<I> {}
...
+                (GeneratorWitness(a_g), GeneratorWitness(b_g)) => a_g.cmp(b_g),
+                (Tuple(a_t), Tuple(b_t)) => a_t.cmp(b_t),
+                (Projection(a_p), Projection(b_p)) => a_p.cmp(b_p),
+                (Opaque(a_d, a_s), Opaque(b_d, b_s)) => a_d.cmp(b_d).then_with(|| a_s.cmp(b_s)),
+                (Param(a_p), Param(b_p)) => a_p.cmp(b_p),
+                (Bound(a_d, a_b), Bound(b_d, b_b)) => a_d.cmp(b_d).then_with(|| a_b.cmp(b_b)),
+                (Placeholder(a_p), Placeholder(b_p)) => a_p.cmp(b_p),
+                (Infer(a_t), Infer(b_t)) => a_t.cmp(b_t),
+                (Error(a_e), Error(b_e)) => a_e.cmp(b_e),
+                _ => Ordering::Equal, // unreachable
```
Comment (oli-obk) [inline compiler/rustc_type_ir/src/sty.rs:384]: same here

**Comment by oli-obk** [comment]
@bors delegate+

r=me with some debug asserts added (they'll save ppl some pain when they add new TyKind variants)

**Comment by WaffleLapkin** [comment]
oops

**Pull request merged**